chromium_src/chrome/browser/bookmarks/android/bookmark_bridge.cc

```diff
@@ -0,0 +1,215 @@
+                     search_engines, favicons),
+      base::BindOnce(&BookmarkBridge::ImportBookmarksImpl,
+                     weak_ptr_factory_.GetWeakPtr()));
+}
+
+void BookmarkBridge::ImportBookmarksImpl(
+    std::tuple<std::vector<ImportedBookmarkEntry>,
+               std::vector<importer::SearchEngineInfo>,
+               favicon_base::FaviconUsageDataList> importedItems) {
+  std::vector<ImportedBookmarkEntry> bookmarks = get<0>(importedItems);
+  std::vector<importer::SearchEngineInfo> search_engines =
+      get<1>(importedItems);
+  favicon_base::FaviconUsageDataList favicons = get<2>(importedItems);
+  auto* writer = new ProfileWriter(profile_);
+
+  if (!bookmarks.empty()) {
+    writer->AddBookmarks(bookmarks, u"Imported");
+  }
+
+  if (!search_engines.empty()) {
+    TemplateURLService::OwnedTemplateURLVector owned_template_urls;
+    for (const auto& search_engine : search_engines) {
+      std::unique_ptr<TemplateURL> owned_template_url = CreateTemplateURL(
+          search_engine.url, search_engine.keyword, search_engine.display_name);
+      if (owned_template_url) {
+        owned_template_urls.push_back(std::move(owned_template_url));
+      }
+    }
+    writer->AddKeywords(std::move(owned_template_urls), false);
+  }
+
+  if (!favicons.empty()) {
+    writer->AddFavicons(favicons);
+  }
+
+  JNIEnv* env = AttachCurrentThread();
+  ScopedJavaLocalRef<jobject> obj =
+      ScopedJavaLocalRef<jobject>(java_bookmark_model_);
+  if (obj.is_null()) {
+    return;
+  }
+
+  Java_BookmarkBridge_bookmarksImported(env, obj, !bookmarks.empty());
+}
+
+std::tuple<std::vector<ImportedBookmarkEntry>,
+           std::vector<importer::SearchEngineInfo>,
+           favicon_base::FaviconUsageDataList>
+BookmarkBridge::ImportBookmarksReader(
+    std::u16string import_path,
+    std::vector<ImportedBookmarkEntry> bookmarks,
+    std::vector<importer::SearchEngineInfo> search_engines,
+    favicon_base::FaviconUsageDataList favicons) {
+  base::FilePath import_path_ = base::FilePath::FromUTF16Unsafe(import_path);
+  bookmark_html_reader::ImportBookmarksFile(
+      base::RepeatingCallback<bool(void)>(),
+      base::BindRepeating(internal::CanImportURL), import_path_, &bookmarks,
+      &search_engines, &favicons);
+
+  return std::make_tuple(bookmarks, search_engines, favicons);
+}
+
+void BookmarkBridge::ExportBookmarks(
+    JNIEnv* env,
+    const JavaParamRef<jobject>& obj,
+    const JavaParamRef<jobject>& java_window,
+    const JavaParamRef<jstring>& j_export_path) {
+  DCHECK(IsLoaded());
+
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+
+  ui::WindowAndroid* window =
+      ui::WindowAndroid::FromJavaWindowAndroid(java_window);
+  CHECK(window);
+
+  std::u16string export_path =
+      base::android::ConvertJavaStringToUTF16(env, j_export_path);
+  base::FilePath export_path_ = base::FilePath::FromUTF16Unsafe(export_path);
+
+  BookmarksExportObserver* observer_ = new FileBookmarksExportObserver(obj);
```

Comment:
I was about  to ask,  why did you choose to allocate raw pointer here and free it in `FileBookmarksExportObserver::OnExportFinished`, but it seems this is the only way because `BookmarkFaviconFetcher::ExecuteWriter` calls `base::ThreadPool::PostTask(...)`.